packages/components/src/components/Calculator/index.tsx

```diff
@@ -0,0 +1,238 @@
+import { PlaygroundSettings } from "../PlaygroundSettings.js";
+import { SqValueWithContext, valueHasContext } from "../../lib/utility.js";
+
+import ReactMarkdown from "react-markdown";
+import {
+  CalculatorState,
+  runSquiggleCode,
+  allFieldValuesAreValid,
+  allFieldResults,
+  initialState,
+} from "./calculatorReducer.js";
+
+import {
+  useViewerContext,
+  useCalculatorFns,
+} from "../SquiggleViewer/ViewerProvider.js";
+
+type Props = {
+  value: SqCalculator;
+  environment: Env;
+  settings: PlaygroundSettings;
+  valueWithContext: SqValueWithContext;
+  renderValue: (
+    value: SqValueWithContext,
+    settings: PlaygroundSettings
+  ) => ReactNode;
+};
+
+const updateFnValue = async (
+  path: SqValuePath,
+  state: CalculatorState,
+  calc: SqCalculator,
+  env: Env,
+  dispatch
+) => {
+  let finalResult: result<SqValue, SqError> | null = null;
+  if (allFieldValuesAreValid(state)) {
+    const results: SqValue[] = allFieldResults(state).map((res) => {
+      if (res && res.ok) {
+        return res.value;
+      } else {
+        throw new Error("Invalid result encountered.");
+      }
+    });
+    finalResult = calc.run(results, env);
+  } else {
+    finalResult = null;
+  }
+
+  dispatch({
+    type: "CALCULATOR_SET_FN_VALUE",
+    payload: { path, value: finalResult },
+  });
+};
+
+async function initStuff(
+  dispatch,
+  path: SqValuePath,
+  state: CalculatorState,
+  calc: SqCalculator,
+  env
+) {
+  for (const name of state.fieldNames) {
+    const field = state.fields[name];
+    const valueResult = await runSquiggleCode(field.code, env);
+    const _action = {
+      type: "CALCULATOR_SET_FIELD_VALUE",
+      payload: { name, value: valueResult, path },
+    };
+    dispatch(_action);
+  }
+  updateFnValue(path, state, calc, env, dispatch);
+}
+
+async function updateCode(
+  dispatch,
+  path: SqValuePath,
+  state: CalculatorState,
+  calc: SqCalculator,
+  env,
+  name: string,
+  code: string,
+  forceUpdate: () => void
+) {
+  const setCodeAction = {
+    type: "CALCULATOR_SET_FIELD_CODE",
+    payload: { path, name: name, code: code },
+  };
+  dispatch(setCodeAction);
+  forceUpdate();
+
+  const valueResult = await runSquiggleCode(code, env);
+  const setValueAction = {
+    type: "CALCULATOR_SET_FIELD_VALUE",
+    payload: { path, name: name, value: valueResult },
+  };
+  dispatch(setValueAction);
+  forceUpdate();
+}
+
+export const Calculator: FC<Props> = ({
+  value,
+  environment,
+  settings,
+  renderValue,
+  valueWithContext,
+}) => {
+  const { path } = valueWithContext.context;
+  const { getSettings, dispatch } = useViewerContext();
+  const [, forceUpdate] = useReducer((x) => x + 1, 0);
+
+  const gotSettings = getSettings({ path, defaults: undefined });
+  const { intitialize } = useCalculatorFns();
+
+  useEffect(() => {
+    const init = async () => {
+      intitialize(path, value);
+      await initStuff(dispatch, path, initialState(value), value, environment);
+      forceUpdate();
```

Comment:
I'm not a fan of these forcedUpdate() methods, copied from the VariableBox. Maybe we could do this with a lower-level component somehow?